core/storage/merkle/polkadot_trie_db/polkadot_codec.cpp

```diff
@@ -0,0 +1,222 @@
+ */
+
+#include "storage/merkle/polkadot_trie_db/polkadot_codec.hpp"
+
+#include "crypto/blake2/blake2s.h"
+#include "storage/merkle/polkadot_trie_db/polkadot_node.hpp"
+
+OUTCOME_CPP_DEFINE_CATEGORY(kagome::storage::merkle, PolkadotCodec::Error, e) {
+  using E = kagome::storage::merkle::PolkadotCodec::Error;
+  switch (e) {
+    case E::SUCCESS:
+      return "success";
+    case E::TOO_MANY_NIBBLES:
+      return "number of nibbles in key is >= 2**16";
+    case E::UNKNOWN_NODE_TYPE:
+      return "unknown polkadot node type";
+  }
+
+  return "unknown";
+}
+
+namespace kagome::storage::merkle {
+
+  inline uint8_t low4nibbles(uint8_t byte) {
+    return byte & 0xFu;
+  }
+
+  inline uint8_t high4nibbles(uint8_t byte) {
+    return (byte >> 4u) & 0xFu;
+  }
+
+  inline uint8_t collectByte(uint8_t low, uint8_t high) {
+    // get low4 from nibbles to avoid check: if(low > 0xff) return error
+    return (low4nibbles(high) << 4u) | low4nibbles(low);
+  }
+
+  inline common::Buffer ushortToBytes(uint16_t b) {
+    common::Buffer out(2, 0);
+    out[0] = b & 0xffu;
+    out[1] = (b >> 8u) & 0xffu;
+    return out;
+  }
+
+  PolkadotCodec::PolkadotCodec(
+      std::shared_ptr<PolkadotCodec::ScaleBufferEncoder> codec)
+      : scale_(std::move(codec)) {}
+
+  common::Buffer PolkadotCodec::nibblesToKey(
+      const common::Buffer &nibbles) const {
+    const auto nibbles_size = nibbles.size();
+    if (nibbles_size == 0) {
+      return {};
+    }
+
+    if (nibbles_size == 1 && nibbles[0] == 0) {
+      return {0};
+    }
+
+    // if nibbles_size is odd, then allocate one more item
+    const size_t size = (nibbles_size + 1) / 2;
+    common::Buffer out(size, 0);
+
+    size_t iterations = size;
+
+    // if number of nibbles is odd, then iterate even number of times
+    bool nimbles_size_odd = nibbles_size % 2 != 0;
+    if (nimbles_size_odd) {
+      --iterations;
+    }
+
+    for (size_t i = 0; i < iterations; ++i) {
+      out[i] = collectByte(nibbles[2 * i], nibbles[2 * i + 1]);
+    }
+
+    // if number of nibbles is odd, then implicitly add 0 as very last nibble
+    if (nimbles_size_odd) {
+      out[iterations] = collectByte(nibbles[2 * iterations], 0);
+    }
+
+    return out;
+  }
+
+  common::Buffer PolkadotCodec::keyToNibbles(const common::Buffer &key) const {
+    auto nibbles = key.size() * 2;
+
+    // if last nibble in `key` is 0
+    bool last_nibble_0 =
+        key.size() > 0 && (high4nibbles(key[key.size() - 1]) == 0);
+    if (last_nibble_0) {
+      --nibbles;
+    }
+
+    common::Buffer out(nibbles, 0);
+    for (size_t i = 0u, size = nibbles / 2; i < size; i++) {
+      out[2 * i] = low4nibbles(key[i]);
+      out[2 * i + 1] = high4nibbles(key[i]);
+    }
+
+    if (last_nibble_0) {
+      out[nibbles] = low4nibbles(key[key.size() - 1]);
+    }
+
+    return out;
+  }
+
+  common::Hash256 PolkadotCodec::hash256(const common::Buffer &buf) const {
+    common::Hash256 out;
+
+    static constexpr uint32_t k256 = 32;
```

Comment:
weird name something like hash_size or k_hash_size would be more readable, or even kHashSize or HASH_SIZE.
not sure what type of constant it is and what case we should use.

Reply:
And by the way, 
```suggestion
    static constexpr uint32_t HASH_SIZE = Hash256::size();
```

Reply:
It is inside the method, e g. "Implementation detail", so naming does not matter much

Reply:
anyway, using magic numbers isn't a good practice, so getting Hash256::size() would be self-explainable.

Reply:
fixed